scripts/pathseq/wdl/pathseq_pipeline.wdl

```diff
@@ -20,179 +26,712 @@
 ## - - one or more read groups all belong to a single sample (SM)
 ##
 ## Output:
-## - BAM file containing microbe-mapped reads and reads of unknown sequence
-## - Tab-separated value (.tsv) file of taxonomic abundance scores
-## - Picard-style metrics files for the filter and scoring phases of the pipeline
+## - non_host_paired_bam, non_host_unpaired_bam: BAM files containing quality-filtered non-host paired and unpaired reads
+## - If filtering_only = false:
+##   - final_bam : BAM file containing microbe-mapped reads and reads of unknown sequence
+##   - taxonomy_scores : tab-separated value (.tsv) file of taxonomic abundance scores
+##   - score_metrics_file : Picard-style metrics file for scoring stage
+##   - non_host_mapped_reads : number of non-host reads mapped to microbe reference
+##   - non_host_unmapped_reads : number of non-host reads that did not map to the microbe reference
+## - If gather_filter_metrics = true:
+##   - filter_metrics_file : Picard-style metrics file for filtering stage
+##   - frac_after_prealigned_filter : estimated fraction of reads remaining after filtering prealigned host reads
+##   - frac_after_qual_cpx_filter : estimated fraction of reads remaining after low-quality and low-complexity filtering
+##   - frac_after_host_filter : estimated fraction of reads remaining after host read filtering
+##   - frac_after_dedup : estimated fraction of reads remaining after deduplication
+##   - frac_final_paired : estimated fraction of reads that were paired in the final output
+##   - frac_final_unpaired : estimated fraction of reads that were unpaired in the final output
+##   - frac_final_total : estimated fraction of reads in the final output
+##   - frac_qual_cpx_filtered : estimated fraction of reads removed by low-quality/low-complexity filtering
+##   - frac_host_filtered : estimated fraction of reads removed by host filtering
+##   - frac_dup_filtered : estimated fraction of reads removed by deduplication
+## - If downsample = true:
+##   - total_reads_if_avail : total number of reads in the original input BAM
 ##
 ########################################################################################################################
 
-task PathseqPipeline {
+workflow PathSeqPipeline {
 
-  # Inputs for this task
   String sample_name
-  File input_bam
+  File input_bam_or_cram
+  # Required if a cram
+  File? input_bam_or_cram_index
```

Comment:
Since WDL error handling is effectively non-existent, I'd rather require an index for bams than code this as optional for both.

Reply:
This is a good point. On the other hand, I had ease-of-use in mind because this way the BAM does not need to be indexed, which is the case for uBAMs generated with most WDLs I've seen in the wild. I could change this to `input_cram_index` to make it less confusing, but the BAM index, when available, speeds up the `Downsample` task.